### PR TITLE
Sam fine tuning example

### PR DESCRIPTION
**Pull request by bnabis93**
Reference
- https://github.com/bowang-lab/MedSAM

### CODE REVIEW

**Comment by bnabis93** [comment]
Point prompt 를 이용한 fine tuning test
1. Point: Bbox center
2. Point: Sampling in mask (Paper)

## How to training the point prompt in original paper
> First, with equal probability either a foreground point or bounding box is selected randomly for the target mask. Points are sampled uniformly from the ground truth mask.

하나의 Mask에서 단 한번의 point를 simulate 하지 않는다. N번의 simulate를 진행한다. 
다음 point는 inference를 한번 진행하여 prediction 과 ground truth의 차이나는 부분인 error region (previous mask prediction and the ground truth mask.) 에서 sampling 한다.

> After making a prediction from this first prompt, subsequent points are selected uniformly from the error region
between the previous mask prediction and the ground truth
mask.

논문에 따르면 많은 정보를 제공하기 위하여 mask의 logit값 또한 prompt에 넣어준다고 한다.

논문에서는 11번의 sampling 과정을 거친다.
- 초기 프롬프트 1, 8번의 sampling, 2번의 반복 (이때는 point 없이 mask만 주는듯) 도합 11번

> We find diminishing returns after 8 iteratively sampled points (we have tested up to 16). Additionally, to encourage the model to benefit from the supplied mask, we also use two more iterations where no additional points are sampled. One of these iterations is randomly inserted among the 8 iteratively sampled points, and the other is always at the end. This gives 11 total iterations: one sampled initial input prompt, 8 iteratively sampled points, and two iterations where no new external information is supplied to the model so it can learn to refine its own mask predictions. We note that using a relatively large number of iterations is possible because our lightweight mask decoder requires less than 1% of the image encoder’s compute and, therefore, each iteration adds only a small overhead. This is unlike previous interactive methods that perform only one or a few interactive steps per optimizer update

---------------------------------------------------

이것만 봐서는 명확한 감이 안온다. 일단은 기존 bbox prompt 대신 point 정보만으로 decoder를 fine tuning 해봐야겠다.

위는 prompt encoder의 학습방법일수도 있다. 지금 내가 원하는것은 decoder의 학습이므로 이부분에 대해서 더 고민해보자.

**Comment by bnabis93** [comment]
## Decoder 학습(Fine-tuning) 결과

### Loss
image

### Validation
```
python validation.py
SAM Dice Coefficient:  0.5904084413572388
MEDSAM Dice Coefficient:  0.8694795289491407
```

### Visualization
gt_bbox_points
contour
